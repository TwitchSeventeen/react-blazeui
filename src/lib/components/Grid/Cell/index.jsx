import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

function Cell(props) {
  const {
    align,
    children,
    className,
    offset,
    width,
    xsmall,
    small,
    medium,
    large,
    xlarge,
    huge,
    noGutter,
    fixedWidth,
    dataName,
    ...other
  } = props;
  const classes = classNames(
    'o-grid__cell',
    { [`o-grid__cell--width-${width}`]: width },
    { [`o-grid__cell--offset-${offset}`]: offset },
    { [`o-grid__cell--${align}`]: align },
    { 'o-grid__cell--no-gutter': noGutter },
    { 'o-grid__cell--width-fixed': fixedWidth },
    { [`o-grid__cell--width-${xsmall}@xsmall`]: xsmall },
    { [`o-grid__cell--width-${small}@small`]: small },
    { [`o-grid__cell--width-${medium}@medium`]: medium },
    { [`o-grid__cell--width-${large}@large`]: large },
    { [`o-grid__cell--width-${xlarge}@xlarge`]: xlarge },
    { [`o-grid__cell--width-${huge}@super`]: huge },
    className,
  );
  return (
    <div
      {...other}
      className={classes}
      data-name={dataName}
    >
      { children }
    </div>
  );
}

Cell.propTypes = {
  /**
   * Vertical alignment of the Cell
   */
  align: PropTypes.oneOf([
    'top',
    'center',
    'bottom',
  ]),
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * The offset of this cell as a percentage of its parent.
   * Must be a multiple of 5, or 33, 66.
   */
  offset: PropTypes.number,
  /**
   * Width of this cell as a percentage of its parent.
   * Must be a multiple of 5, or 33, 66.
   */
  width: PropTypes.number,
  /**
   * Cell display for xsmall screens
   * Must be a multiple of 5, or 33, 66.
   */
  xsmall: PropTypes.number,
  /**
   * Cell display for small screens
   * Must be a multiple of 5, or 33, 66.
   */
  small: PropTypes.number,
  /**
   * Cell display for medium screens
   * Must be a multiple of 5, or 33, 66.
   */
  medium: PropTypes.number,
  /**
   * Cell display for large screens
   * Must be a multiple of 5, or 33, 66.
   */
  large: PropTypes.number,
  /**
   * Cell display for xlarge screens
   * Must be a multiple of 5, or 33, 66.
   */
  xlarge: PropTypes.number,
  /**
   * Cell display for super screens
   * Must be a multiple of 5, or 33, 66.
   */
  huge: PropTypes.number,
  /**
   * Specifies no spacing for the Cell
   */
  noGutter: PropTypes.bool,
  /**
   * Specifies this Cell has a fixed width
   */
  fixedWidth: PropTypes.bool,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
};

Cell.defaultProps = {
  align: null,
  children: null,
  className: null,
  offset: null,
  width: null,
  xsmall: null,
  small: null,
  medium: null,
  large: null,
  xlarge: null,
  huge: null,
  noGutter: false,
  fixedWidth: false,
  dataName: 'grid-cell',
};

export default Cell;
