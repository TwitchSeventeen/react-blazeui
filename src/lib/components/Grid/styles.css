.o-grid--demo .o-grid__cell {
    border: 2px solid #00bfff;
    background-color: #b2f5b5;
}