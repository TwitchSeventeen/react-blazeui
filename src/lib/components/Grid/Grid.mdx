---
name: Grid
route: /objects/grid
menu: Objects
---

import { Playground, PropsTable } from 'docz'
import Grid from './index';

# Grid

The grid is the easiest way to produce a fluid responsive layout. Its easy to use and very flexible.

Colour key:
<ul>
    <li class="key key-grid">grids are pink</li>
    <li class="key key-cell">cells are blue</li>
    <li class="key key-padding">padding is green</li>
</ul>

## Basic Grid
Use the Grid component to wrap your Grid.Cell components
<Playground>
    <Grid className="demo">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Widths
Use the **width** prop to apply widths to your cells. If you add a width to one cell the other cells will automatically take up the rest of the space equally.
<Playground>
    <Grid className="demo">
        <Grid.Cell width={50}>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

Use the **full** prop on the Grid to make all the cells full width.
<Playground>
    <Grid full className="demo">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Offsets
To "push" a cell horizontally across the Grid use the **offset** prop. Use in-conjunction with **widths** to perfectly align your cells.

<Playground>
    <Grid className="demo">
        <Grid.Cell width={20}>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell width={20} offset={60}>
            <div className="text">second</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Wrapping
A grid is essentially a row of cells. However cells can wrap onto the next line within the Grid.

Here we have 2 cells, 50% and 66.6666% wide, but the Grid is only 100% wide...eek!

<Playground>
    <Grid className="demo">
        <Grid.Cell width={50}>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell width={66}>
            <div className="text">second</div>
        </Grid.Cell>
    </Grid>
</Playground>

To solve this we can wrap the cells within the Grid by adding the **wrap** prop.

<Playground>
    <Grid wrap className="demo">
        <Grid.Cell width={50}>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell width={66}>
            <div className="text">second</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Nesting
You can nest grids inside cells with relative ease...with nesting you're into advanced territory!

<Playground>
    <Grid className="demo">
        <Grid.Cell>
            <Grid>
                <Grid.Cell>
                    <div className="text">first</div>
                </Grid.Cell>
                <Grid.Cell>
                    <div className="text">second</div>
                </Grid.Cell>
            </Grid>
            <Grid>
                <Grid.Cell>
                    <div className="text">third</div>
                </Grid.Cell>
                <Grid.Cell>
                    <div className="text">fourth</div>
                </Grid.Cell>
            </Grid>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">fifth</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Vertical Alignment
### Per Grid
If one of your Cells is tall you can add the **align** prop to the Grid.

_top_, _center_, and _bottom_ are supported.

<Playground>
    <Grid className="demo" align="bottom">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">fourth
            <br/>fourth
            <br/>fourth
            <br/>fourth
            <br/>fourth
            <br/>
            </div>
        </Grid.Cell>
    </Grid>
</Playground>

### Per Cell
You can override the vertical alignment of all the cells in the grid by specifying per cell.

_top_, _center_, and _bottom_ are supported.

<Playground>
    <Grid className="demo">
        <Grid.Cell align="top">
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell align="center">
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell align="bottom">
            <div className="text">third</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">fourth
            <br/>fourth
            <br/>fourth
            <br/>fourth
            <br/>fourth
            <br/>
            </div>
        </Grid.Cell>
    </Grid>
</Playground>

## Responsiveness
The Grid can be instructed to go full width at certain break points to maximise the screen space and help produce great user experiences on all sorts of devices.

The Grid has props for each breakpoint that take a value of either _'fit'_ or _'full'_.


**NOTE**: To avoid conflicting with JavaScript, the Blaze UI _super_ breakpoint has been renamed to _huge_.

<Playground>
    <Grid className="demo" small="full" medium="fit" large="full">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

The above example will display the Cells full width for small and large size screens. For medium screens the Cells will adjust to fit the available space. Try it out, either resize your browser or rotate your phone into landscape.

Because they use viewport media queries, simply resizing the example box won't work.

## Responsive Cells
As well as Grid level responsiveness you also have control over the Cells within a Grid.

Individual Cells can change width at specific break points by using props.

<Playground>
    <Grid wrap className="demo">
        <Grid.Cell width={100} small={33}>
            <div className="text">small</div>
        </Grid.Cell>
        <Grid.Cell width={100} medium={33}>
            <div className="text">medium</div>
        </Grid.Cell>
        <Grid.Cell width={100} large={33}>
            <div className="text">large</div>
        </Grid.Cell>
    </Grid>
</Playground>

As the viewport increases in size the Cells go from 100% wide to 33% wide one after another, rather than all at once.

The props apply to the screen sizes _**and up**_, in a mobile first manner. In other words:

Set Cell width to 33% when screen width is greater than or equal to _xsmall_ | _small_ | _medium_ | _large_ | _xlarge_ | _huge_

## Gutters
Grid comes with a gutter around each grid Cell; some padding either side.

Sometimes you want your Cells to be flush with each other. To turn off all gutters use the Grid **noGutter** prop.

<Playground>
    <Grid className="demo" noGutter>
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

...or Cell specific **noGutter**.

<Playground>
    <Grid className="demo">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell>
            <div className="text">second</div>
        </Grid.Cell>
        <Grid.Cell noGutter>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Fixed widths
Usually Cells in a Grid will be _**fluid and responsive**_. If however you want to set a Cell to a fixed width add the **fixedWidth** prop and give it a specific width.

<Playground>
    <Grid className="demo">
        <Grid.Cell>
            <div className="text">first</div>
        </Grid.Cell>
        <Grid.Cell fixedWidth style={{ width: '200px' }}>
            <div className="text">static width of 200px</div>
        </Grid.Cell>
        <Grid.Cell noGutter>
            <div className="text">third</div>
        </Grid.Cell>
    </Grid>
</Playground>

## Properties
### Grid
<PropsTable of={Grid} />

### Cell
<PropsTable of={Grid.Cell} />