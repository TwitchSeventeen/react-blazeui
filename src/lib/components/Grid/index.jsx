import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Cell from './Cell';

function Grid (props) {
  const { align, children, className, full, wrap, xsmall, small, medium, large, xlarge, huge, noGutter } = props;
  const classes = classNames(
      'o-grid',
      { ['o-grid--full']: full },
      { ['o-grid--wrap']: wrap },
      { [`o-grid--${align}`]: align },
      { ['o-grid--no-gutter']: noGutter },
      { [`o-grid--xsmall-${xsmall}`]: xsmall },
      { [`o-grid--small-${small}`]: small },
      { [`o-grid--medium-${medium}`]: medium },
      { [`o-grid--large-${large}`]: large },
      { [`o-grid--xlarge-${xlarge}`]: xlarge },
      { [`o-grid--super-${huge}`]: huge },
      className
  );
  return (
      <div className={classes}>
        { children }
      </div>
  );
}

Grid.propTypes = {
  /**
   * Vertical alignment of the Grid contents
   */
  align: PropTypes.oneOf([
      'top',
      'center',
      'bottom',
  ]),
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Makes all child Cells full width
   */
  full: PropTypes.bool,
  /**
   * Wrap child Cells within the Grid
   */
  wrap: PropTypes.bool,
  /**
   * Cell display for xsmall screens
   */
  xsmall: PropTypes.oneOf([
      'fit',
      'full',
  ]),
  /**
   * Cell display for small screens
   */
  small: PropTypes.oneOf([
    'fit',
    'full',
  ]),
  /**
   * Cell display for medium screens
   */
  medium: PropTypes.oneOf([
    'fit',
    'full',
  ]),
  /**
   * Cell display for large screens
   */
  large: PropTypes.oneOf([
    'fit',
    'full',
  ]),
  /**
   * Cell display for xlarge screens
   */
  xlarge: PropTypes.oneOf([
    'fit',
    'full',
  ]),
  /**
   * Cell display for huge screens
   */
  huge: PropTypes.oneOf([
    'fit',
    'full',
  ]),
  /**
   * Specifies no spacing between Cells
   */
  noGutter: PropTypes.bool,
};

Grid.defaultProps = {
  align: null,
  children: null,
  className: null,
  full: false,
  wrap: false,
  xsmall: null,
  small: null,
  medium: null,
  large: null,
  xlarge: null,
  huge: null,
  noGutter: false,
};

Grid.Cell = Cell;

export default Grid;