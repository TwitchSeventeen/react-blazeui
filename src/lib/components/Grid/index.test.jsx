import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Grid from './index';

describe('Grid', () => {
  it('should be defined', () => {
    expect(Grid).toBeDefined();
  });

  it('renders correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Grid>lorem</Grid>);

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with full width children', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Grid full>
        <Grid.Cell>lorem</Grid.Cell>
        <Grid.Cell>ipsum</Grid.Cell>
      </Grid>
  );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with wrapped children', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
        <Grid wrap>
          <Grid.Cell width={50}>lorem</Grid.Cell>
          <Grid.Cell width={66}>ipsum</Grid.Cell>
        </Grid>
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with responsive props', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
        <Grid
          xsmall="fit"
          small="full"
          medium="fit"
          large="full"
          xlarge="fit"
          huge="full"
        >
          <Grid.Cell width={50}>lorem</Grid.Cell>
          <Grid.Cell width={50}>ipsum</Grid.Cell>
        </Grid>
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without gutters', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
        <Grid noGutter>
          <Grid.Cell width={50}>lorem</Grid.Cell>
          <Grid.Cell width={50}>ipsum</Grid.Cell>
        </Grid>
    );
  });
});