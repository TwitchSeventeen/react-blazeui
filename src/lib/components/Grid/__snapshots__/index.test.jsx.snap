// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
<div
  className="o-grid u-text"
>
  lorem
</div>
`;

exports[`Grid renders correctly with full width children 1`] = `
<div
  className="o-grid u-text o-grid--full"
>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={null}
    xlarge={null}
    xsmall={null}
  >
    lorem
  </Cell>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={null}
    xlarge={null}
    xsmall={null}
  >
    ipsum
  </Cell>
</div>
`;

exports[`Grid renders correctly with responsive props 1`] = `
<div
  className="o-grid u-text o-grid--xsmall-fit o-grid--small-full o-grid--medium-fit o-grid--large-full o-grid--xlarge-fit o-grid--super-full"
>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={50}
    xlarge={null}
    xsmall={null}
  >
    lorem
  </Cell>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={50}
    xlarge={null}
    xsmall={null}
  >
    ipsum
  </Cell>
</div>
`;

exports[`Grid renders correctly with wrapped children 1`] = `
<div
  className="o-grid u-text o-grid--wrap"
>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={50}
    xlarge={null}
    xsmall={null}
  >
    lorem
  </Cell>
  <Cell
    align={null}
    className={null}
    fixedWidth={false}
    huge={null}
    large={null}
    medium={null}
    noGutter={false}
    offset={null}
    small={null}
    width={66}
    xlarge={null}
    xsmall={null}
  >
    ipsum
  </Cell>
</div>
`;
